From 84884bbc21a63b61d698138f549baae152efd878 Mon Sep 17 00:00:00 2001
From: Aaron Allen <aaron@flipstone.com>
Date: Mon, 24 May 2021 20:13:52 -0500
Subject: [PATCH] Increase purview of --dry-run and --only-download flags
 (#7379)

Changes the behavior of the following commands under the `--dry-run` and
`--only-download` flags
- `v2-configure` and `v2-freeze` do not write their respective files
- `v2-exec` and `v2-run` do not run the target executable

Adds package tests for the configure and freeze commands.
---
 .../src/Distribution/Client/CmdConfigure.hs   | 51 ++++++++++++-------
 .../src/Distribution/Client/CmdExec.hs        |  9 +++-
 .../src/Distribution/Client/CmdFreeze.hs      | 17 ++++---
 .../src/Distribution/Client/CmdRun.hs         | 26 ++++++----
 .../src/Distribution/Client/ProjectConfig.hs  |  1 +
 .../NewConfigure/ConfigFile/ConfigFile.cabal  |  7 +++
 .../NewConfigure/ConfigFile/Setup.hs          |  2 +
 .../NewConfigure/ConfigFile/cabal.out         |  5 ++
 .../NewConfigure/ConfigFile/cabal.project     |  1 +
 .../NewConfigure/ConfigFile/cabal.test.hs     | 18 +++++++
 .../NewFreeze/FreezeFile/new_freeze.out       |  4 ++
 .../NewFreeze/FreezeFile/new_freeze.test.hs   |  5 ++
 changelog.d/pr-7407                           |  4 ++
 13 files changed, 115 insertions(+), 35 deletions(-)
 create mode 100644 cabal-testsuite/PackageTests/NewConfigure/ConfigFile/ConfigFile.cabal
 create mode 100644 cabal-testsuite/PackageTests/NewConfigure/ConfigFile/Setup.hs
 create mode 100644 cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.out
 create mode 100644 cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.project
 create mode 100644 cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.test.hs
 create mode 100644 changelog.d/pr-7407

diff --git a/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/ConfigFile.cabal b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/ConfigFile.cabal
new file mode 100644
index 000000000..3173c99e8
--- /dev/null
+++ b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/ConfigFile.cabal
@@ -0,0 +1,7 @@
+name:                ConfigFile
+version:             0.1.0.0
+author:              Foo Bar
+maintainer:          cabal-dev@haskell.org
+build-type:          Simple
+cabal-version:       >=1.10
+
diff --git a/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/Setup.hs b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/Setup.hs
new file mode 100644
index 000000000..9a994af67
--- /dev/null
+++ b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/Setup.hs
@@ -0,0 +1,2 @@
+import Distribution.Simple
+main = defaultMain
diff --git a/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.out b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.out
new file mode 100644
index 000000000..f26ed6aac
--- /dev/null
+++ b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.out
@@ -0,0 +1,5 @@
+# cabal v2-configure
+Config file not written due to flag(s).
+# cabal v2-configure
+Config file not written due to flag(s).
+# cabal v2-configure
diff --git a/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.project b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.project
new file mode 100644
index 000000000..e6fdbadb4
--- /dev/null
+++ b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.project
@@ -0,0 +1 @@
+packages: .
diff --git a/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.test.hs b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.test.hs
new file mode 100644
index 000000000..bc0a57473
--- /dev/null
+++ b/cabal-testsuite/PackageTests/NewConfigure/ConfigFile/cabal.test.hs
@@ -0,0 +1,18 @@
+import Test.Cabal.Prelude
+
+-- Test that 'cabal v2-configure' generates the config file appropriately
+main = withShorterPathForNewBuildStore $ \storeDir ->
+  cabalTest . withSourceCopy $ do
+    cwd <- fmap testCurrentDir getTestEnv
+    let configFile = cwd </> "cabal.project.local"
+
+    shouldNotExist configFile
+
+    -- should not create config file with --dry-run or --only-download
+    cabalG ["--store-dir=" ++ storeDir] "v2-configure" ["--dry-run"]
+    cabalG ["--store-dir=" ++ storeDir] "v2-configure" ["--only-download"]
+    shouldNotExist configFile
+
+    -- should create the config file
+    cabalG ["--store-dir=" ++ storeDir] "v2-configure" []
+    shouldExist configFile
diff --git a/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.out b/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.out
index be5731382..4b2d614c9 100644
--- a/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.out
+++ b/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.out
@@ -7,6 +7,10 @@ In order, the following would be built:
  - my-library-dep-2.0 (lib) (requires build)
  - my-local-package-1.0 (exe:my-exe) (first run)
 # cabal v2-freeze
+Freeze file not written due to flag(s)
+# cabal v2-freeze
+Freeze file not written due to flag(s)
+# cabal v2-freeze
 Resolving dependencies...
 Wrote freeze file: <ROOT>/new_freeze.dist/source/cabal.project.freeze
 # cabal v2-build
diff --git a/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.test.hs b/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.test.hs
index e2a3d5988..912649bba 100644
--- a/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.test.hs
+++ b/cabal-testsuite/PackageTests/NewFreeze/FreezeFile/new_freeze.test.hs
@@ -16,6 +16,11 @@ main = withShorterPathForNewBuildStore $ \storeDir ->
       -- v2-build should choose the latest version for the dependency.
       cabalG' ["--store-dir=" ++ storeDir] "v2-build" ["--dry-run"] >>= assertUsesLatestDependency
 
+      -- should not create freeze file with --dry-run or --only-download flags
+      cabalG' ["--store-dir=" ++ storeDir] "v2-freeze" ["--dry-run"]
+      cabalG' ["--store-dir=" ++ storeDir] "v2-freeze" ["--only-download"]
+      shouldNotExist freezeFile
+
       -- Freeze a dependency on the older version.
       cabalG ["--store-dir=" ++ storeDir] "v2-freeze" ["--constraint=my-library-dep==1.0"]
 
diff --git a/changelog.d/pr-7407 b/changelog.d/pr-7407
new file mode 100644
index 000000000..5991d3d84
--- /dev/null
+++ b/changelog.d/pr-7407
@@ -0,0 +1,4 @@
+synopsis: --dry-run and --only-download effect v2-configure, v2-freeze, v2-run, and v2-exec
+pr: #7407
+issues: #7379
+decription: { v2-configure, v2-freeze, v2-run, and v2-exec now behave expectedly under the --dry-run and --only-download flags }
-- 
2.25.1

